File: Homework2/Hw2.Tests/CalculatorTests.cs
using Hw2;
using Xunit;

namespace Hw2Tests
{
    public class CalculatorTests
    {
        [Theory]
        [InlineData(15, 5, CalculatorOperation.Plus, 20)]
        [InlineData(15, 5, CalculatorOperation.Minus, 10)]
        [InlineData(15, 5, CalculatorOperation.Multiply, 75)]
        [InlineData(15, 5, CalculatorOperation.Divide, 3)]
        public void TestAllOperations(int value1, int value2, CalculatorOperation operation, int expectedValue)
        {
            var res = Calculator.Calculate(value1, value2, operation);

            Assert.Equal(expectedValue, res);
        }

        [Fact]
        public void TestInvalidOperation()
        {
            Assert.Throws<InvalidOperationException>(
                () => Calculator.Calculate(1, 2, CalculatorOperation.Undefined));
        }

        [Fact]
        public void TestDividingNonZeroByZero()
        {
            Assert.Throws<DivideByZeroException>(
                () => Calculator.Calculate(1, 0, CalculatorOperation.Divide));
        }

        [Fact]
        public void TestDividingZeroByNonZero()
        {
            var res = Calculator.Calculate(0, 1, CalculatorOperation.Divide);

            Assert.Equal(0, res);
        }

        [Fact]
        public void TestDividingZeroByZero()
        {
            Assert.Throws<DivideByZeroException>(
                () => Calculator.Calculate(1, 0, CalculatorOperation.Divide));
        }
    }
}

File: Homework5/Hw5.Tests/CalculatorTests.fs
module Hw5Tests.CalculatorTests

open Hw5
open Hw5.Calculator
open Microsoft.FSharp.Core
open Xunit

let epsilon: decimal = 0.001m
        
let intTest: obj [] list = [
    [| 15; 5; CalculatorOperation.Plus; 20 |]
    [| 15; 5; CalculatorOperation.Minus; 10 |]
    [| 15; 5; CalculatorOperation.Multiply; 75 |]
    [| 15; 5; CalculatorOperation.Divide; 3 |]
]
        
[<Theory>]
[<MemberData(nameof(intTest))>]
let ``+, -, *, / work return correct calculation results with ints`` (value1 : int, value2: int, operation, expectedValue : int) =
    //act
    let actual = calculate value1 operation value2
    
    //assert
    Assert.Equal(expectedValue, actual)

let doubleTest: obj [] list = [
    [| 15.6; 5.6; CalculatorOperation.Plus; 21.2 |]
    [| 15.6; 5.6; CalculatorOperation.Minus; 10 |]
    [| 15.6; 5.6; CalculatorOperation.Multiply; 87.36 |]
    [| 15.6; 5.6; CalculatorOperation.Divide; 2.7857 |]
]

[<Theory>]
[<MemberData(nameof(doubleTest))>]
let ``+, -, *, / work return correct calculation results with floats``
    (value1 : float, value2: float, operation, expectedValue : float) =
    //act
    let actual = abs (expectedValue - calculate value1 operation value2)
        
    //assert
    Assert.True(actual |> decimal < epsilon)
    
let decimalTest: obj [] list = [
    [| 15.6m; 5.6m; CalculatorOperation.Plus; 21.2m |]
    [| 15.6m; 5.6m; CalculatorOperation.Minus; 10m |]
    [| 15.6m; 5.6m; CalculatorOperation.Multiply; 87.36m |]
    [| 15.6m; 5.6m; CalculatorOperation.Divide; 2.7857m |]
]

[<Theory>]
[<MemberData(nameof(decimalTest))>]
let ``+, -, *, / work return correct calculation results with decimals``
    (value1 : decimal, value2: decimal, operation, expectedValue : decimal) =
    //act
    let actual = abs (expectedValue - calculate value1 operation value2)
        
    //assert
    Assert.True(actual |> decimal < epsilon)
    